core/llm_env.go

```diff
@@ -29,140 +37,198 @@ type LLMTool struct {
+		objsByName:        map[string]dagql.Object{},
+		objsByID:          map[string]dagql.Object{},
+		varsByName:        map[string]string{},
+		typeCount:         map[string]int{},
+		idByHash:          map[digest.Digest]string{},
+		functionMask:      map[string]bool{},
+		needsToolHint:     endpoint.Provider == Google,
+		needsSystemPrompt: endpoint.Provider == Google,
 	}
 }
 
+//go:embed llm_dagger_prompt.md
+var defaultSystemPrompt string
+
+func (env *LLMEnv) DefaultSystemPrompt() string {
+	if env.needsSystemPrompt {
+		return defaultSystemPrompt
+	}
+	return ""
+}
+
 func (env *LLMEnv) Clone() *LLMEnv {
 	cp := *env
-	cp.history = cloneSlice(cp.history)
-	cp.vars = cloneMap(cp.vars)
-	cp.objsByHash = cloneMap(cp.objsByHash)
+	cp.objsByName = cloneMap(cp.objsByName)
+	cp.typeCount = cloneMap(cp.typeCount)
+	cp.idByHash = cloneMap(cp.idByHash)
+	cp.objsByID = cloneMap(cp.objsByID)
 	return &cp
 }
 
-// Lookup dagql typedef for a given dagql value
-func (env *LLMEnv) typedef(srv *dagql.Server, val dagql.Typed) *ast.Definition {
-	return srv.Schema().Types[val.Type().Name()]
-}
-
 // Return the current selection
-func (env *LLMEnv) Current() dagql.Typed {
-	if len(env.history) == 0 {
-		return nil
-	}
-	return env.history[len(env.history)-1]
+func (env *LLMEnv) Current() dagql.Object {
+	return env.current
 }
 
-func (env *LLMEnv) With(val dagql.Typed) {
-	env.history = append(env.history, val)
+func (env *LLMEnv) Select(obj dagql.Object, functions ...string) {
+	env.current = obj
+	clear(env.functionMask)
+	for _, fn := range functions {
+		env.functionMask[fn] = true
+	}
 }
 
 // Save a value at the given key
-func (env *LLMEnv) Set(key string, value dagql.Typed) string {
-	prev := env.vars[key]
-	env.vars[key] = value
-	if obj, ok := dagql.UnwrapAs[dagql.Object](value); ok {
-		env.objsByHash[obj.ID().Digest()] = value
-	}
-	if prev != nil {
-		return fmt.Sprintf("The variable %q has changed from %s to %s.", key, env.describe(prev), env.describe(value))
-	}
-	return fmt.Sprintf("The variable %q has been set to %s.", key, env.describe(value))
+func (env *LLMEnv) Set(key string, obj dagql.Object) {
+	env.objsByName[key] = obj
+	env.intern(obj)
 }
 
-// Get a value saved at the given key
-func (env *LLMEnv) Get(key string) (dagql.Typed, error) {
-	if val, exists := env.vars[key]; exists {
-		return val, nil
-	}
-	if _, hash, ok := strings.Cut(key, "@"); ok {
-		// strip Type@ prefix if present
-		// TODO: figure out the best place to do this
-		key = hash
+// Get an object saved at a given key
+func (env *LLMEnv) GetObject(key, expectedType string) (dagql.Object, error) {
+	if expectedType != "" {
+		// for maximal LLM compatibility, assume type for numeric ID args
+		if onlyNum, err := strconv.Atoi(key); err == nil {
+			key = fmt.Sprintf("%s#%d", expectedType, onlyNum)
+		}
 	}
-	if val, exists := env.objsByHash[digest.Digest(key)]; exists {
+	// strip $foo var prefix
+	key = strings.TrimPrefix(key, "$")
```

Comment:
What does this do?

Reply:
Mostly a remnant of when it was all variable-oriented, since some models (or even the same model) would opt to include the `$` prefix when referencing variables. The general idea is to be liberal in what we accept, considering how moody a lot of models can be, and a lot of them aren't capable of strictly adhering to a format.